Sources/AuthgearError.swift

```diff
@@ -24,12 +24,88 @@ public enum AuthgearError: Error {
     case unauthenticatedUser
     case publicKeyNotFound
     case error(Error)
+
+    public static var errorDomain: String { "AuthgearError" }
+    public var errorCode: Int {
+        switch self {
+        case .cancel:
+            return 0
+        case .unexpectedHttpStatusCode:
+            return 1
+        case .serverError:
+            return 2
+        case .oauthError:
+            return 3
+        case .anonymousUserNotFound:
+            return 4
+        case .biometricPrivateKeyNotFound:
+            return 5
+        case .biometricNotSupportedOrPermissionDenied:
+            return 6
+        case .biometricNoPasscode:
+            return 7
+        case .biometricNoEnrollment:
+            return 8
+        case .biometricLockout:
+            return 9
+        case .cannotReauthenticate:
+            return 10
+        case .invalidJWT:
+            return 11
+        case .unauthenticatedUser:
+            return 11
+        case .publicKeyNotFound:
+            return 11
+        case .error:
+            return 12
+        }
+    }
+
+    public var errorUserInfo: [String: Any] {
+        var info: [String: Any] = [:]
+        switch self {
+        case let .unexpectedHttpStatusCode(code, _):
+            info["code"] = code
+        case let .serverError(err):
+            info = err.errorUserInfo
+        case let .oauthError(err):
+            info = err.errorUserInfo
+        case let .biometricNotSupportedOrPermissionDenied(err):
+            info["error"] = err.localizedDescription
+        case let .biometricNoPasscode(err):
+            info["error"] = err.localizedDescription
+        case let .biometricNoEnrollment(err):
+            info["error"] = err.localizedDescription
+        case let .biometricLockout(err):
+            info["error"] = err.localizedDescription
+        case let .invalidJWT(jwt):
+            info["jwt"] = jwt
+        case let .error(err):
+            info["error"] = err.localizedDescription
```

Comment:
For cases that wrap an `Error`, I think we should cast the `Error` to `NSError`, and then construct the error info as follows:

```
{
  "domain": NSError.domain,
  "code": NSError.code,
  "info": NSError.info,
  "localizedDescription": NSError.localizedDescription
}
```

That also means we should make all our error structs/classes (including ServerError, OAuthError, and LatteError) implements CustomNSError too.

I would like to know now that AuthgearError implements CustomNSError, if we cast it to NSError, would the domain, code, and info correct?

Reply:
@louischan-oursky Yes. Actually, all `Error` have a default implementation of `NSError`, which use its class name domain, and the enumerator as code. However we cannot control the code and have no idea how to map it back to the exact enum. Implementing CustomNSError let us customize this behavior.